File: frontend/src/components/azure-blob-selector/azure-blob-storage-service.ts
/*
 * Copyright sovity GmbH and/or licensed to sovity GmbH under one or
 * more contributor license agreements. You may not use this file except
 * in compliance with the "Elastic License 2.0".
 *
 * SPDX-License-Identifier: Elastic-2.0
 */
import {BlobServiceClient} from '@azure/storage-blob';

class AzureBlobStorageService {
  public static async getContainers(
    sasToken: string,
    storageAccountName: string,
  ): Promise<string[]> {
    const sasUrl =
      'https://' + storageAccountName + '.blob.core.windows.net?' + sasToken;
    const blobServiceClient = new BlobServiceClient(sasUrl);

    const containerNames: string[] = [];
    for await (const containerItem of blobServiceClient.listContainers()) {
      containerNames.push(containerItem.name);
    }
    return containerNames;
  }

  public static async getBlobs(
    sasToken: string,
    containerName: string,
    storageAccountName: string,
  ): Promise<string[]> {
    const containerUrl =
      'https://' +
      storageAccountName +
      '.blob.core.windows.net/' +
      containerName +
      '?' +
      sasToken;
    const blobServiceClient = new BlobServiceClient(containerUrl);
    const containerClient = blobServiceClient.getContainerClient('');

    const blobList = [];
    for await (const blob of containerClient.listBlobsFlat()) {
      blobList.push(blob.name);
    }

    return blobList;
  }
}

export default AzureBlobStorageService;

File: connector/ce/connector-ce/src/main/kotlin/de/sovity/edc/ce/api/ui/pages/blob_storage/AzureBlobStorageService.kt
/*
 * Copyright sovity GmbH and/or licensed to sovity GmbH under one or
 * more contributor license agreements. You may not use this file except
 * in compliance with the "Elastic License 2.0".
 *
 * SPDX-License-Identifier: Elastic-2.0
 */
package de.sovity.edc.ce.api.ui.pages.blob_storage

import de.sovity.edc.ce.api.ui.model.AzureStorageListBlobsRequest
import de.sovity.edc.ce.api.ui.model.AzureStorageListContainersRequest
import de.sovity.edc.runtime.simple_di.Service

@Service
class AzureBlobStorageService(
    private val blobStoreApi: SovityBlobStoreApiImpl,
) {

    fun listContainerNames(request: AzureStorageListContainersRequest): List<String> =
        blobStoreApi.listAllContainers(
            accountKey = request.storageAccountVaultKey,
            accountName = request.storageAccountName
        ).map { it.name }

    fun listBlobNames(request: AzureStorageListBlobsRequest): List<String> =
        blobStoreApi.listContainer(
            accountKey = request.storageAccountVaultKey,
            accountName = request.storageAccountName,
            containerName = request.containerName
        ).map { it.name }
}
